### PR TITLE
fileutil: add Windows flock

### PR DESCRIPTION
**Pull request by gyuho**
Fix https://github.com/coreos/etcd/issues/1972.
Related to https://github.com/coreos/etcd/issues/2515.

### CODE REVIEW

**Comment by gyuho** [comment]
Tested in Windows environment. Needed some test skipping but the core function
works (`PurgeFile` with flock). Implementation is based on:

- https://github.com/golang/build/blob/master/cmd/builder/filemutex_windows.go
- https://github.com/boltdb/bolt/blob/master/bolt_windows.go

/cc @heyitsanthony @xiang90

File: pkg/fileutil/lock_windows.go [hunk 1]
```diff
@@ -33,39 +52,128 @@ type Lock interface {
...
+			err = syscall.EINVAL
+		}
+	}
+	return
 }
 
-// TryLock acquires exclusivity on the lock without blocking
-func (l *lock) TryLock() error {
-	return nil
+func unlockFileEx(h syscall.Handle, reserved, locklow, lockhigh uint32, ol *syscall.Overlapped) (err error) {
```
Comment (heyitsanthony) [inline pkg/fileutil/lock_windows.go]: drop the reserved argument since it should always be 0?

Comment (gyuho) [inline pkg/fileutil/lock_windows.go]: Ok will do. Thanks!

File: pkg/fileutil/lock_windows.go [hunk 2]
```diff
@@ -33,39 +52,128 @@ type Lock interface {
...
+	if atomic.LoadInt32(&fm.locked) == 1 {
+		return ErrLocked
+	}
+	atomic.StoreInt32(&fm.locked, 1)
+	fm.mu.Lock()
+	var flag uint32 = LOCKFILE_EXCLUSIVE_LOCK
+	if flags != 0 {
+		flag |= flags
+	}
+	if fm.fd != syscall.InvalidHandle {
```
Comment (heyitsanthony) [inline pkg/fileutil/lock_windows.go]: `if fm.fd == syscall.InvalidHandle { return nil }`?

Comment (gyuho) [inline pkg/fileutil/lock_windows.go]: Fixed. Thanks!

File: pkg/fileutil/lock_windows.go [hunk 3]
```diff
@@ -33,39 +52,128 @@ type Lock interface {
...
+	fm.mu.Unlock()
+	atomic.StoreInt32(&fm.locked, 0)
 	return nil
 }
 
-// Unlock unlocks the lock
-func (l *lock) Unlock() error {
+func (fm *fileMutex) lockFile(flags uint32) error {
+	if atomic.LoadInt32(&fm.locked) == 1 {
+		return ErrLocked
```
Comment (heyitsanthony) [inline pkg/fileutil/lock_windows.go]: should this happen in the `Lock()` case? I think it's supposed to block?

File: pkg/fileutil/lock_windows.go [hunk 4]
```diff
@@ -33,39 +52,131 @@ type Lock interface {
...
+	}
+	atomic.StoreInt32(&fm.locked, 1)
+	fm.mu.Lock()
```
Comment (heyitsanthony) [inline pkg/fileutil/lock_windows.go]: I think this is missing a few matching `Unlock`s in the error cases

File: pkg/fileutil/lock_windows.go (hunk 3 above)
Comment (gyuho) [inline pkg/fileutil/lock_windows.go]: Right. I don't think it's needed. Will delete and just let mutex handle the synchronization.

File: pkg/fileutil/purge_test.go [hunk 5]
```diff
@@ -20,22 +20,29 @@ import (
...
 	"path"
 	"reflect"
+	"runtime"
 	"testing"
 	"time"
 )
 
 func TestPurgeFile(t *testing.T) {
-	dir, err := ioutil.TempDir("", "purgefile")
+	wd, err := os.Getwd()
```
Comment (heyitsanthony) [inline pkg/fileutil/purge_test.go]: windows can't do tempdir right?

Comment (gyuho) [inline pkg/fileutil/purge_test.go]: Yeah doesn't work in windows. Added comments.

File: pkg/fileutil/lock_windows.go [hunk 6]
```diff
@@ -33,39 +52,131 @@ type Lock interface {
...
+	}
+	var (
+		t       time.Time
+		timeout = time.Second
+	)
+	for {
+		// https://github.com/boltdb/bolt/blob/master/bolt_windows.go
+		// If we're beyond our timeout then return an error.
+		// This can only occur after we've attempted a flock once.
+		if t.IsZero() {
```
Comment (heyitsanthony) [inline pkg/fileutil/lock_windows.go]: I don't think Lock needs to care about timeouts?

File: pkg/fileutil/lock_windows.go (hunk 4 above)
Comment (gyuho) [inline pkg/fileutil/lock_windows.go]: > I think this is missing a few matching Unlocks in the error cases

Which one are we talking about? Could you explain? Thanks!

File: pkg/fileutil/lock_windows.go (hunk 6 above)
Comment (gyuho) [inline pkg/fileutil/lock_windows.go]: Right. boltdb used it for db connection. Just deleted. Thanks.

File: pkg/fileutil/lock_windows.go (hunk 4 above)
Comment (heyitsanthony) [inline pkg/fileutil/lock_windows.go]: if `lockFileEx` gives `err != nil`, `lockFile` will exit with an error still holding the mutex; if something else tries to `Lock`, there'll be a deadlock

**Comment by gyuho** [comment]
@heyitsanthony 

Actually it needs to retry with timeout... In tests, I see the flock fails at first try and but succeeds in second try.

And added lock releasing in case lockFile errors.

PTAL

Thanks.

**Comment by heyitsanthony** [comment]
@gyuho I don't think it needs the timeout since the other Lock interface implementations don't have a timeout. The retry is fine.

**Comment by gyuho** [comment]
@heyitsanthony I see. Misunderstood your comment. I removed the `timeout` and just keep the retrying. PTAL. Thanks!

**Comment by gyuho** [comment]
@heyitsanthony Sorry I was wrong.

It was an error from something else. Fixed to remove `os.Getwd` calls.
Tested in Windows and passed without `Getwd`.

Thanks!

**Comment by gyuho** [comment]
This gets 

```
etcdserver: create wal error: write infra2.etcd/member/wal/0000000000000000-0000000000000000.wal: The process cannot access the file because another process has locked a portion of the file.
```

Please ignore this for now.

**Comment by heyitsanthony** [comment]
code looks ok but I feel the tests should pass instead of skip before this is good to go

**Comment by gyuho** [comment]
@heyitsanthony Agree. I will try to remove all window specific exceptions. Will test-run cluster as well. Thanks.

**Comment by gyuho** [comment]
Windows locks a file when it opens. Even though this correctly flocks, it cannot write:

```
package main

import (
	"fmt"
	"log"
	"os"

	"github.com/coreos/etcd/pkg/fileutil"
)

func main() {
	f, err := os.OpenFile("my_file.txt", os.O_WRONLY|os.O_APPEND|os.O_CREATE, 0600)
	if err != nil {
		log.Fatal(err)
	}
	defer os.Remove("my_file.txt")

	l, err := fileutil.NewLock("my_file.txt")
	if err != nil {
		log.Fatal(err)
	}
	err = l.Lock()
	fmt.Println(l, err)

	_, err = f.WriteString("aaaaa")
	fmt.Println(err)
	// The process cannot access the file because another process has locked a portion of the file.
}

```

This will not work (same use case for `wal` package in etcd).
So `syscall.Flock` is not portable in the same way as unix. I will try other ways.

Thanks.

**Pull request closed without merging**